### DISCUSSION TITLE
issue with long running custom commands

### DISCUSSION THREAD

**Discussion by rienafairefr** [starship/starship#3196] [Q&A]
I've tried to use starship to show some info about the current directory context (the currently activated pulumi stack to be more precise). Works fine

But the command takes a bit of time (not very long). if I'm in a directory where the custom command fires, and if I fire ctrl-C twice (e.g. cancelling a command, but then ctrl C once more, probably some cargo cult thing, doing ctrl-c multiple times to be sure you killed it lol. Anyway (AFAIK), if ctrl-c is fired during the time that the custom command runs, then the whole shell crashes. If I ctrl-c fast in another directory where the custom command does not fire, no problems

Example (the P symbol with `prod` is the output from the custom command ) of what I see:
```
…/infra/deploy on  master [$✘!?] via  v12.22.7 P prod took 2s 
❯ ^C
^C

Warning: Program '/bin/bash' crashed.
```
I'm tempted to say it might not be the fault of starship, might be a misconfiguration of my bash or whatever, why I didn't push it as a bug report. Anyone ever experienced something like this, any workaround so that I can ctrl-c like a mad man ^^ ?

**Comment by chipbuster** [comment]
This is interesting. I can't replicate this in bash, using various terminals like Terminator or WezTerm. I just get an infinite series of Ctrl-Cs, but it doesn't crash (and crucially, it doesn't seem to stop the slow command either).

The command I used to get "slow" performance was `sleep 2`. Perhaps this is somehow linked to Pulumi?

**Reply by rienafairefr** [reply]
Maybe it's something of how I've configured starship/the shell. Let me see if we can find the culprit :
I'm running these tests in konsole on kde plasma, bash --versions gives `GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)`
This is the full starship.toml : 
```
[directory]
truncation_length = 8
truncate_to_repo = false

[custom.pulumi]
command = "pulumi stack | head -1 | cut -d' ' -f4 | sed 's/.$//'"
style = "bold 105"
format = "[$symbol$output]($style) "
symbol = "P "
when = "[[ -f Pulumi.yaml ]]"
``` 
When replacing command with "sleep 2", I also get the `Warning: Program '/bin/bash' crashed.` when Ctrl+C, so I don't think it's a weird thing about pulumi's signal handling. 

Replaced the sleep 2 with a sleep 3600, looked at ps --foret:
```
matthieu    8355  0.5  0.2 388604 91592 ?        Sl   09:49   0:03  |   \_ /usr/bin/konsole
matthieu   14986  0.0  0.0  13532  5244 pts/2    Ss+  09:58   0:00  |   |   \_ /bin/bash
matthieu   15017  0.0  0.0  40976  2172 pts/2    Sl+  09:58   0:00  |   |       \_ /home/matthieu/.local/bin/starship prompt --status=0 --pipestatus 0 --jobs=0 --cmd-duration=2
matthieu   15035  0.0  0.0  12048  3360 pts/2    S+   09:58   0:00  |   |           \_ /bin/bash
matthieu   15036  0.0  0.0  10632   592 pts/2    S+   09:58   0:00  |   |               \_ sleep 3600
```
nothing really surprising there I guess ? Maybe the command_duration module is it ?

~/.bashrc contains `eval "$(starship init bash)"`

if it helps, strace on the /bin/bash process while I ctrl-c:

```
❯ sudo strace -p 15035
strace: Process 15035 attached
wait4(-1, 0x7ffff8d4cdb0, 0, NULL)      = ? ERESTARTSYS (To be restarted if SA_RESTART is set)
--- SIGINT {si_signo=SIGINT, si_code=SI_KERNEL} ---
rt_sigreturn({mask=[CHLD]})             = -1 EINTR (Interrupted system call)
wait4(-1, [{WIFSIGNALED(s) && WTERMSIG(s) == SIGINT}], 0, NULL) = 15036
--- SIGHUP {si_signo=SIGHUP, si_code=SI_KERNEL} ---
+++ killed by SIGHUP +++
```

another strace, on the "sleep" process this time
```
❯ sudo strace -p 15609
strace: Process 15609 attached
restart_syscall(<... resuming interrupted read ...>) = ? ERESTART_RESTARTBLOCK (Interrupted by signal)
--- SIGINT {si_signo=SIGINT, si_code=SI_KERNEL} ---
+++ killed by SIGINT +++
```

I can ctrl-c a long running command in a normal shell and get back the shell prompt
```
~ 
❯ sleep 30
^C

~ 
❯
```
and that /bin/bash strace is different
```
❯ sudo strace -p 14901
strace: Process 14901 attached
wait4(-1, [{WIFSIGNALED(s) && WTERMSIG(s) == SIGINT}], WSTOPPED|WCONTINUED, NULL) = 16568
rt_sigprocmask(SIG_BLOCK, [CHLD TSTP TTIN TTOU], [CHLD], 8) = 0
ioctl(255, TIOCSPGRP, [14901])          = 0
rt_sigprocmask(SIG_SETMASK, [CHLD], NULL, 8) = 0
ioctl(255, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(255, SNDCTL_TMR_STOP or TCSETSW, {B38400 opost isig icanon echo ...}) = 0
ioctl(255, TCGETS, {B38400 opost isig icanon echo ...}) = 0
ioctl(255, TIOCGWINSZ, {ws_row=19, ws_col=131, ws_xpixel=0, ws_ypixel=0}) = 0
write(1, "\n", 1)
```
Hopefully this helps.
I tried to understand the content of starship bash init (`starship init bash --print-full-init`) but I'm not sure where to start

Looked at the advanced section with preexec/precmd thing (https://starship.rs/advanced-config/#custom-pre-prompt-and-pre-execution-commands-in-bash) and setup my .bashrc thus

```
function precmd(){
    echo "precmd"
}

starship_precmd_user_func="precmd"

function preexec(){
    echo "preexec"
}
trap preexec DEBUG
eval "$(starship init bash)"
```

This is the content of the shell output when I open a new terminal emulator window, navigate to a directory where custom.pulumi triggers (running `sleep 2`), then ctrl-c

```
preexec
preexec
precmd

~ 
❯ cd Projects/test-long-pulumi/
precmd
^C

Warning: Program '/bin/bash' crashed.
```

**Reply by chipbuster** [reply]
I'm sorry I didn't respond---life got very, very busy right around that time.

I'm trying this with `GNU bash, version 5.0.17(1)-release (x86_64-pc-linux-gnu)` (got it from the Arch Linux Archives) and not getting the same behavior as you. I've even tried your custom trapping, and I see the same deal as you up until I try to Ctrl+C the process, at which point...

```
preexec
preexec
precmd

~ 
❯ cd ~/tmp/
precmd
^C
precmd
^C
precmd
^C
precmd

~/tmp via   P  took 2s 
❯ 
```

So something clearly differs between our setups.

If you are able to, could you run an invocation of `bash -x` and `bash -x -l` and post the results here? This will reveal the contents of your shell configurations *and* your system shell configurations, so don't do it if you have sensitive stuff in there (or at least go through and erase them).

I'm also wondering if our shells were compiled differently (different compile-time options leads to different runtime behaviours, but it's hard to tell how a shell was compiled). Could you tell me what distro and version you're using? If I spin up a new VM there and try this and get your results, I'll have a reproduction platform. If I don't, then we can at least be pretty sure that it's something local to your system (config, most likely, but other possibilitieS)

**Reply by chipbuster** [reply]
Just as a note: the SIGHUP that your child bash gets is *very* suspicious, but I don't really know how to probe into that further remotely. It almost smells like a signal is being sent to the wrong place, or to the wrong recipient, but I don't understand how that's happening.
